src/service/history-impl.js

```diff
@@ -34,6 +34,21 @@ const HISTORY_PROP_ = 'AMP.History';
 /** @typedef {number} */
 let HistoryIdDef;
 
+/** @typedef {{
+  stackIndex: number,
+  title: string,
+  fragment: string,
+  ampBindState: (JsonObject|undefined)
+}} */
+let HistoryStateDef;
+
+/** @typedef {{
+  title: (string|undefined),
+  fragment: (string|undefined),
+  ampBindState: (JsonObject|undefined)
+}} */
+let HistoryStateUpdateDef;
```

Comment:
Optional: Might be nicer to have this as:

```js
/** @typedef {{title: (string|undefined), fragment: (string|undefined), data: (JsonObject|undefined)}} */
let HistoryStateDef;

/** @typedef {{index: HistoryIdDef, state: HistoryStateDef}} */
let HistoryStackFrameDef;
```

Don't want to make you change all the tests though.